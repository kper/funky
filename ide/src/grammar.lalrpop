use std::str::FromStr;
use crate::ir::ast::*;

grammar;

Comma<T>: Vec<T> = { 
    <v:(<T> ",")*> <e:T?> => match e {
        None => v,
        Some(e) => {
            let mut v = v;
            v.push(e);
            v
        }
    }
};

pub Program : Program = {
   <functions:Function*> => Program {
       functions: functions
   } 
};

pub Function : Function = {
    "define" <name:Id> "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: Vec::new(),
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Id> "(param" <reg:Reg*> ")" "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: reg,
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Name> "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: Vec::new(),
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    },
    "define" <name:Name> "(param" <reg:Reg*> ")" "(result" <result:Num> ")" "(define" <def:Reg*> ")" "{" <instr:Instruction*>  "}" ";" => {
        Function {
            name,
            params: reg,
            definitions: def,
            results_len: result as usize,
            instructions: instr
        }
    }
};

pub Instruction : Instruction = {
    "BLOCK" <n:Id> => Instruction::Block(n),
    <dest:Reg> "=" <src:Reg> => Instruction::Assign(dest, src),
    <dest:Reg> "=" <n:Num> => Instruction::Const(dest, n),
    <dest:Reg> "=" "UNKNOWN" => Instruction::Unknown(dest),
    <dest:Reg> "=" "op" <src:Reg> => Instruction::Unop(dest, src),
    <dest:Reg> "=" <src1:Reg> "op" <src2:Reg> => Instruction::BinOp(dest, src1, src2),
    <dest:Reg> "=" "phi" <src1:Reg> <src2:Reg> => Instruction::Phi(dest, src1, src2),
    <dest:Reg> "=" "STORE" <src:Reg> "TO" <offset:Reg> "ALIGN" <align:Num> <size:Num> => Instruction::Store,
    "GOTO" <n:Id> => Instruction::Jump(n),
    "IF" <reg:Reg> "THEN" "GOTO" <n:Id> "ELSE" "GOTO" <n2:Id> => {
        Instruction::Conditional(reg, vec![n, n2])
    },
    "IF" <reg:Reg> "THEN" "GOTO" <n:Id> => {
        Instruction::Conditional(reg, vec![n])
    },
    "TABLE GOTO" <n:Id*> "ELSE" "GOTO" <n2:Id> => {
        let mut n = n.clone();
        n.push(n2);
        Instruction::Table(n)
    },
    "KILL" <reg:Reg> => Instruction::Kill(reg),
    "CALL" <n:Name> "(" <r:Reg*> ")" => Instruction::Call(n, r, vec![]),
    <dest:Reg*> "<-" "CALL" <n:Name> "(" <r:Reg*> ")" => Instruction::Call(n, r, dest),
    "CALL" <n:Id> "(" <r:Reg*> ")" => Instruction::Call(n, r, vec![]),
    <dest:Reg*> "<-" "CALL" <n:Id> "(" <r:Reg*> ")" => Instruction::Call(n, r, dest),
    "RETURN" <n:Reg*> ";" => {
        Instruction::Return(n)
    },
    "CALL INDIRECT" <n:Id*> "(" <p:Reg*> ")" => Instruction::CallIndirect(n, p, vec![]),
    <dest:Reg*> "<-" "CALL INDIRECT" <n:Id*> "(" <p:Reg*> ")" => Instruction::CallIndirect(n, p, dest),
};

Id : String = {
    <n:Num> => (n as usize).to_string()
};

pub Reg: String = {
    "%" <n:Num> => format!("%{}", n),
    "%" <n:Name> => format!("%{}", n),
};

pub Num: f64 = <s:r"[+-]?[0-9]*[.,]?[0-9]+"> => f64::from_str(s).unwrap();
pub Name: String = <s:r"[a-zA-Z]+"> => s.to_string();