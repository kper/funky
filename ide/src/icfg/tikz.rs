use crate::icfg::graph2::{Edge, Graph};
use log::debug;

pub const TAU: usize = 1;

pub fn render_to(graph: &Graph) -> String {
    let mut str_vars = String::new();

    let mut index = 0;
    let mut functions: Vec<_> = graph.functions.iter().map(|(_, f)| f).collect();

    functions.sort_by(|a, b| b.name.cmp(&a.name));

    for function in functions {
        let function_name = &function.name;
        debug!("Drawing function {}", function_name);

        let facts = graph.facts.iter().filter(|x| &x.function == function_name);
        let notes = graph.notes.iter().filter(|x| &x.function == function_name);

        for fact in facts {
            debug!("Drawing fact");

            str_vars.push_str(&format!(
                "\\node[circle,fill,inner sep=1pt,label=left:{}] ({}) at ({}, {}) {{}};\n",
                fact.belongs_to_var.replace("%", "\\%"),
                fact.id,
                index + fact.track,
                fact.pc,
            ));
        }

        for note in notes {
            debug!("Drawing note");

            str_vars.push_str(&format!(
                "\\node[font=\\tiny] (note_{}) at ({}, {}) {{{}}};\n",
                note.id,
                index as f64 - 1.5,   //x
                note.pc as f64 - 0.5, //y
                note.note.replace("%", "").replace("\"", "").escape_debug(),
            ));
        }

        index += function.definitions + TAU + 1;
    }

    for edge in graph.edges.iter() {
        match edge {
            Edge::Normal { from, to , curved } => {
                if *curved {
                    str_vars.push_str(&format!("\t\t\\path[->, bend right] ({}) edge ({});\n", from.id, to.id));
                }
                else {
                    str_vars.push_str(&format!("\t\t\\path[->] ({}) edge ({});\n", from.id, to.id));
                }
            }
            Edge::Call { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, green] ({}) [bend left] edge node {{ }} ({});\n",
                    from.id, to.id
                ));
            }
            Edge::CallToReturn { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->] ({}) edge node {{ }} ({});\n",
                    from.id, to.id
                ));
            }
            Edge::Return { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, red] ({}) [bend right] edge  node {{ }} ({});\n",
                    from.id, to.id
                ));
            }
        }
    }

    template(str_vars)
}

fn template(inject: String) -> String {
    format!(
        "   \\documentclass{{article}}
            \\usepackage{{pgf, tikz}}
            \\usetikzlibrary{{arrows, automata}}
            \\begin{{document}}
                \\begin{{tikzpicture}}
                    [node distance=3cm]
                    \\tikzstyle{{every state}}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    {}

                \\end{{tikzpicture}}
            \\end{{document}}",
        inject
    )
}
