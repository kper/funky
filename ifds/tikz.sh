#!/bin/bash

rm /tmp/graph.tex
rm /tmp/graph.pdf
tail -n +6 $1 > /tmp/graph.tex
xelatex /tmp/graph.tex && evince graph.pdf
