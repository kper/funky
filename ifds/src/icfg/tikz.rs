use crate::counter::Counter;
use crate::icfg::graph::{Edge, Fact, Graph};
use crate::icfg::state::State;
use log::debug;
use std::cmp::Ordering;

const TAU: usize = 1;

struct FactWrapper<'a> {
    id: usize,
    fact: &'a Fact,
}

impl<'a> FactWrapper<'a> {
    pub fn get(&'a self) -> &'a Fact {
        self.fact
    }
}

/// Convert the given graph into string in .tex format.
pub fn render_to(graph: &Graph, state: &State) -> String {
    let mut str_vars = String::new();

    let mut index = 0;
    let mut functions: Vec<_> = state.functions.iter().map(|(_, f)| f).collect();

    functions.sort_by(|a, b| b.name.cmp(&a.name));

    let mut counter = Counter::default();

    let mut facts: Vec<_> = graph
        .edges
        .iter()
        .map(|x| x.get_from())
        .chain(graph.edges.iter().map(|x| x.to()))
        .collect();

    fn chain_ordering(o1: Ordering, o2: Ordering) -> Ordering {
        match o1 {
            Ordering::Equal => o1,
            _ => o2,
        }
    }

    facts.sort_by(|a, b| {
        chain_ordering(
            chain_ordering(
                b.function.cmp(&a.function),
                b.belongs_to_var.cmp(&a.belongs_to_var),
            ),
            b.next_pc.cmp(&a.next_pc),
        )
    });
    facts.dedup();

    let facts = facts
        .into_iter()
        .map(|x| FactWrapper {
            id: counter.get(),
            fact: x,
        })
        .collect::<Vec<_>>();

    for function in functions {
        let function_name = &function.name;
        debug!("Drawing function {}", function_name);

        let facts = facts.iter().filter(|x| &x.get().function == function_name);

        let notes = state.notes.iter().filter(|x| &x.function == function_name);

        let mut vars : Vec<_> = facts.clone().map(|x| &x.get().belongs_to_var).collect();
        vars.sort_by(|a, b| {
            b.cmp(a)
        });
        vars.dedup();

        for fact in facts {
            debug!("Drawing fact");

            str_vars.push_str(&format!(
                "\\node[circle,fill,inner sep=1pt,label=left:${}$] ({}) at ({}, {}) {{ }};\n",
                fact.get().belongs_to_var.replace("%", "\\%"),
                fact.id,
                index + vars.iter().position(|x| x == &&fact.get().belongs_to_var).unwrap(),
                fact.get().next_pc,
            ));
        }

        for note in notes {
            debug!("Drawing note");

            str_vars.push_str(&format!(
                "\\node[font=\\tiny] (note_{}) at ({}, {}) {{{}}};\n",
                note.id,
                index as f64 - 1.5,   //x
                note.pc as f64 + 0.5, //y
                note.note.replace("%", "").replace("\"", "").escape_debug(),
            ));
        }

        index += function.definitions + TAU + 1;
    }

    let get_fact_id = |fact| {
        let pos = facts.iter().position(|x| x.get() == fact).unwrap();
        facts.get(pos).unwrap().id
    };

    for edge in graph.edges.iter() {
        match edge {
            Edge::Normal { from, to, curved } => {
                if *curved {
                    str_vars.push_str(&format!(
                        "\t\t\\path[->, bend right] ({}) edge ({});\n",
                        get_fact_id(from),
                        get_fact_id(to)
                    ));
                } else {
                    str_vars.push_str(&format!(
                        "\t\t\\path[->] ({}) edge ({});\n",
                        get_fact_id(from),
                        get_fact_id(to)
                    ));
                }
            }
            Edge::Call { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, green] ({}) [bend left] edge node {{ }} ({});\n",
                    get_fact_id(from),
                    get_fact_id(to)
                ));
            }
            Edge::CallToReturn { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->] ({}) edge node {{ }} ({});\n",
                    get_fact_id(from),
                    get_fact_id(to)
                ));
            }
            Edge::Return { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, red] ({}) [bend right] edge  node {{ }} ({});\n",
                    get_fact_id(from),
                    get_fact_id(to)
                ));
            }
            Edge::Path { from, to } => {
                if from != to {
                    str_vars.push_str(&format!(
                        "\t\t\\path[->, blue] ({}) [bend right] edge  node {{ }} ({});\n",
                        get_fact_id(from),
                        get_fact_id(to)
                    ));
                } else {
                    str_vars.push_str(&format!(
                        "\t\t\\path[->, blue] ({}) [loop right] edge  node {{ }} ({});\n",
                        get_fact_id(from),
                        get_fact_id(to)
                    ));
                }
            }
            Edge::Summary { from, to } => {
                str_vars.push_str(&format!(
                    "\t\t\\path[->, red] ({}) [bend left] edge  node {{ }} ({});\n",
                    get_fact_id(from),
                    get_fact_id(to)
                ));
            }
        }
    }

    template(str_vars)
}

fn template(inject: String) -> String {
    format!(
        "   \\documentclass{{standalone}}
            \\usepackage{{pgf, tikz}}
            \\usetikzlibrary{{arrows, automata}}
            \\begin{{document}}
                \\begin{{tikzpicture}}
                    [node distance=3cm]
                    \\tikzstyle{{every state}}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    {}

                \\end{{tikzpicture}}
            \\end{{document}}",
        inject
    )
}
