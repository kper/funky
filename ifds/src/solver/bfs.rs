use crate::icfg::graph::{Fact, Graph};
use crate::icfg::state::State;
use crate::solver::*;
use std::collections::VecDeque;

use log::debug;

pub struct Bfs;

impl GraphReachability for Bfs {
    fn all_sinks(&mut self, graph: &mut Graph, state: &State, req: &Request) -> Vec<Taint> {
        let mut queue: VecDeque<&Fact> = VecDeque::new();
        let mut seen = Vec::new();

        if let Some(start) = state
            .get_facts_at(&req.function, req.pc)
            .unwrap()
            .find(|x| &x.belongs_to_var == req.variable.as_ref().unwrap())
        {
            debug!("Adding start {:#?}", start);
            queue.push_back(start);
        } else {
            return Vec::new();
        }

        while let Some(node) = queue.pop_front() {
            debug!("Popping node {:?}", node);
            seen.push(node);
            for child in graph
                .edges
                .iter()
                .filter(|x| x.get_from() == node)
                .map(|x| x.to())
            {
                //debug!("Adding child {:#?}", graph.query_by_fact_id(child));

                if !seen.contains(&child) {
                    debug!("queue child {:?}", child);
                    queue.push_back(child);
                }
            }
        }

        seen.into_iter()
            .map(|x| Taint {
                variable: x.belongs_to_var.clone(),
                pc: x.next_pc,
                function: x.function.clone(),
            })
            .collect()
    }
}
