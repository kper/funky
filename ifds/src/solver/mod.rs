//! The implementation for extracting the taints from the graph.

use crate::icfg::graph::{Edge, Graph};
use anyhow::{Context, Result};
use std::collections::HashSet;

type PC = usize;

pub struct IfdsSolver;

/// Represents the taints from the execution of the solver.
#[derive(Debug)]
pub struct Taint {
    pub variable: String,
    pub function: String,
    pub pc: PC,
}

/// Request is the datastructure which defines at what instruction the [`Graph`] should be build from.
#[derive(Debug, Clone)]
pub struct Request {
    pub variable: Option<String>,
    pub function: String,
    pub pc: PC,
}

pub trait Solver {
    /// Get all taints for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Result<Vec<Taint>>;

    /// Check if the statement at `req` and the statement `resp` have a taint.
    fn is_taint(&mut self, graph: &mut Graph, req: &Request, resp: &Request) -> Result<bool>;

    /// Return all tainted variables names at any statement for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn sinks_var(&mut self, graph: &mut Graph, req: &Request) -> Result<HashSet<String>>;
}

impl Solver for IfdsSolver {
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Result<Vec<Taint>> {
        assert!(req.variable.is_some());
        assert!(req.variable.as_ref().unwrap().starts_with("%"));

        let function = &req.function;

        let taints: Vec<_> = graph
            .edges
            .iter()
            .filter(|x| {
                matches!(x, Edge::Path { .. })
                    && &x.to().function == function
                    //&& x.get_from().var_is_taut
                    && x.get_from().next_pc == req.pc
            })
            .map(|x| x.to())
            .filter(|x| !x.var_is_taut)
            .map(|x| Taint {
                function: x.function.clone(),
                pc: x.next_pc.checked_sub(1).unwrap_or(0),
                variable: x.belongs_to_var.clone(),
            })
            .collect();

        Ok(taints)
    }

    fn sinks_var(&mut self, graph: &mut Graph, req: &Request) -> Result<HashSet<String>> {
        assert!(req.variable.is_some());
        assert!(req.variable.as_ref().unwrap().starts_with("%"));

        let function = &req.function;

        let taints = graph
            .edges
            .iter()
            .filter(|x| {
                matches!(x, Edge::Path { .. })
                    && &x.to().function == function
                    && x.get_from().next_pc == req.pc
            })
            .map(|x| x.to())
            .filter(|x| !x.var_is_taut)
            .map(|x| x.belongs_to_var.clone())
            .collect::<HashSet<_>>();

        Ok(taints)
    }

    fn is_taint(&mut self, graph: &mut Graph, req: &Request, resp: &Request) -> Result<bool> {
        let all_sinks = self
            .all_sinks(graph, req)
            .context("Cannot compute the sinks")?;

        let var = resp
            .variable
            .as_ref()
            .context("Please specify a variable for the response")?;

        Ok(all_sinks
            .into_iter()
            .find(|x| x.function == resp.function && x.pc == resp.pc && &x.variable == var)
            .is_some())
    }
}

pub trait GraphReachability {
    /// Get all taints for the graph. The edges must be already computed.
    /// The result depends on the `req` [`Request`]. Only the requested function and instruction which have
    /// an higher or equal program counter.
    fn all_sinks(&mut self, graph: &mut Graph, req: &Request) -> Vec<Taint>;
}
