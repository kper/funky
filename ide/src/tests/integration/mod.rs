use crate::ir::wasm_ast::IR;
use funky::engine::module::ModuleInstance;
use funky::engine::*;
use insta::assert_snapshot;
use validation::validate;
use wasm_parser::{parse, read_wasm};

use crate::grammar::*;
use crate::icfg::convert2::ConvertSummary;
use crate::icfg::tikz::render_to;
use crate::solver::Request;

macro_rules! wasm {
    ($input:expr) => {{
        // Read it
        let file = read_wasm!($input);
        let module = parse(file).expect("Parsing failed");
        assert!(validate(&module).is_ok());

        let imports = Vec::new();

        let instance = ModuleInstance::new(&module);
        let engine = Engine::new(
            instance,
            &module,
            Box::new(funky::debugger::RelativeProgramCounter::default()),
            &imports,
        )
        .unwrap();

        let mut ir = IR::new();

        ir.visit(&engine).unwrap();

        ir
    }};
}

macro_rules! run {
    ($name:expr, $req:expr, $fs:expr) => {
        let ir = wasm!($fs);

        let ir_code = ir.buffer();

        let mut convert = ConvertSummary::new();

        let prog = ProgramParser::new().parse(&ir_code).unwrap();

        let graph = convert.visit(&prog, &$req).unwrap();

        let output = render_to(&graph);

        assert_snapshot!(format!("{}", $name), output);
    };
}

#[test]
fn test_add() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("add", req, "../tests/add.wasm");
}

#[test]
fn test_fib() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("fib", req, "../tests/fib.wasm");
}

#[test]
fn test_fac() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("fac", req, "../tests/fac.wasm");
}

#[test]
fn test_br_table() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("br_table", req, "../tests/br_table.wasm");
}
