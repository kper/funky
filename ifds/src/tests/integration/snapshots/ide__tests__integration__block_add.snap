---
source: src/tests/integration/mod.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:\%4] (36) at (5, 10) { };
\node[circle,fill,inner sep=1pt,label=left:\%4] (35) at (5, 9) { };
\node[circle,fill,inner sep=1pt,label=left:\%4] (34) at (5, 8) { };
\node[circle,fill,inner sep=1pt,label=left:\%4] (33) at (5, 7) { };
\node[circle,fill,inner sep=1pt,label=left:\%2] (32) at (3, 6) { };
\node[circle,fill,inner sep=1pt,label=left:\%4] (31) at (5, 6) { };
\node[circle,fill,inner sep=1pt,label=left:\%2] (30) at (3, 5) { };
\node[circle,fill,inner sep=1pt,label=left:\%4] (29) at (5, 5) { };
\node[circle,fill,inner sep=1pt,label=left:\%2] (28) at (3, 4) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (27) at (0, 10) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (26) at (0, 9) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (25) at (0, 8) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (24) at (0, 7) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (23) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (22) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (21) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (20) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (19) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:\%2] (18) at (3, 3) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (16) at (0, 10) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (15) at (0, 9) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (14) at (0, 8) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (13) at (0, 7) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (12) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (11) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (10) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (9) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (8) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (7) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (6) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:taut] (0) at (0, 2) { };
\node[font=\tiny] (note_0) at (-1.5, 0.5) {Block(0)};
\node[font=\tiny] (note_1) at (-1.5, 1.5) {Block(1)};
\node[font=\tiny] (note_2) at (-1.5, 2.5) {Assign(2, 0)};
\node[font=\tiny] (note_3) at (-1.5, 3.5) {Assign(3, 1)};
\node[font=\tiny] (note_4) at (-1.5, 4.5) {BinOp(4, 3, 2)};
\node[font=\tiny] (note_5) at (-1.5, 5.5) {Kill(3)};
\node[font=\tiny] (note_6) at (-1.5, 6.5) {Kill(2)};
\node[font=\tiny] (note_7) at (-1.5, 7.5) {Jump(2)};
\node[font=\tiny] (note_8) at (-1.5, 8.5) {Block(2)};
\node[font=\tiny] (note_9) at (-1.5, 9.5) {Return([4])};
		\path[->, blue] (0) [loop right] edge  node { } (0);
		\path[->, blue] (0) [bend right] edge  node { } (6);
		\path[->, blue] (0) [bend right] edge  node { } (7);
		\path[->, blue] (0) [bend right] edge  node { } (9);
		\path[->, blue] (0) [bend right] edge  node { } (10);
		\path[->, blue] (0) [bend right] edge  node { } (11);
		\path[->, blue] (0) [bend right] edge  node { } (12);
		\path[->, blue] (0) [bend right] edge  node { } (13);
		\path[->, blue] (0) [bend right] edge  node { } (14);
		\path[->, blue] (0) [bend right] edge  node { } (15);
		\path[->, blue] (0) [bend right] edge  node { } (16);
		\path[->, blue] (0) [bend right] edge  node { } (18);
		\path[->, blue] (0) [bend right] edge  node { } (28);
		\path[->, blue] (0) [bend right] edge  node { } (29);
		\path[->, blue] (0) [bend right] edge  node { } (30);
		\path[->, blue] (0) [bend right] edge  node { } (31);
		\path[->, blue] (0) [bend right] edge  node { } (32);
		\path[->, blue] (0) [bend right] edge  node { } (33);
		\path[->, blue] (0) [bend right] edge  node { } (34);
		\path[->, blue] (0) [bend right] edge  node { } (35);
		\path[->, blue] (0) [bend right] edge  node { } (36);
		\path[->] (0) edge (6);
		\path[->] (6) edge (7);
		\path[->] (7) edge (8);
		\path[->] (8) edge (9);
		\path[->] (9) edge (10);
		\path[->] (10) edge (11);
		\path[->] (11) edge (12);
		\path[->] (12) edge (13);
		\path[->] (13) edge (14);
		\path[->] (14) edge (15);
		\path[->] (15) edge (16);
		\path[->] (0) edge (18);
		\path[->] (0) edge (19);
		\path[->] (6) edge (20);
		\path[->] (7) edge (21);
		\path[->] (9) edge (22);
		\path[->] (11) edge (23);
		\path[->] (12) edge (24);
		\path[->, bend right] (13) edge (25);
		\path[->] (14) edge (26);
		\path[->] (15) edge (27);
		\path[->] (18) edge (28);
		\path[->] (28) edge (29);
		\path[->] (28) edge (30);
		\path[->] (29) edge (31);
		\path[->] (30) edge (32);
		\path[->] (31) edge (33);
		\path[->, bend right] (33) edge (34);
		\path[->] (34) edge (35);
		\path[->] (35) edge (36);


                \end{tikzpicture}
            \end{document}
